posthog/models/test/test_scheduled_change.py

```diff
@@ -0,0 +1,144 @@
+import json
+from posthog.models import ScheduledChange
+from posthog.test.base import BaseTest
+
+
+class TestScheduledChange(BaseTest):
+    def setUp(self):
+        super().setUp()
+        self.scheduled_change = ScheduledChange.objects.create(
+            team=self.team,
+            record_id="test-flag",
+            model_name=ScheduledChange.AllowedModels.FEATURE_FLAG,
+            payload={"operation": "update_status", "value": True},
+            scheduled_at="2025-07-22 12:00:00+00:00",
+            created_by=self.user,
+        )
+
+    def test_formatted_failure_reason_with_null(self):
+        """Test that null failure_reason returns 'Unknown error'"""
+        self.scheduled_change.failure_reason = None
+        self.assertEqual(self.scheduled_change.formatted_failure_reason, "Unknown error")
+
+    def test_formatted_failure_reason_with_empty_string(self):
+        """Test that empty failure_reason returns 'Unknown error'"""
+        self.scheduled_change.failure_reason = ""
+        self.assertEqual(self.scheduled_change.formatted_failure_reason, "Unknown error")
+
+    def test_formatted_failure_reason_legacy_string(self):
+        """Test that legacy string failure reasons are returned as-is"""
+        self.scheduled_change.failure_reason = "Database connection failed"
+        self.assertEqual(self.scheduled_change.formatted_failure_reason, "Database connection failed")
+
+    def test_formatted_failure_reason_json_unrecoverable(self):
+        """Test JSON format for unrecoverable errors"""
+        failure_context = {
+            "error": "Invalid payload structure",
+            "error_type": "ValidationError",
+            "error_classification": "unrecoverable",
+            "will_retry": False,
+            "retry_count": 1,
+            "max_retries": 5,
+            "timestamp": "2025-07-22T22:47:00Z",
+            "hostname": "worker-1",  # Sensitive info that should not be exposed
+            "task_id": "abc123",  # Sensitive info that should not be exposed
+        }
+        self.scheduled_change.failure_reason = json.dumps(failure_context)
+
+        expected = "Invalid payload structure (permanent error)"
+        self.assertEqual(self.scheduled_change.formatted_failure_reason, expected)
+
+    def test_formatted_failure_reason_json_retry_exhausted(self):
+        """Test JSON format for retry exhausted"""
+        failure_context = {
+            "error": "Database connection lost",
+            "error_type": "OperationalError",
+            "error_classification": "recoverable",
+            "will_retry": False,
+            "retry_exhausted": True,
+            "retry_count": 5,
+            "max_retries": 5,
+            "timestamp": "2025-07-22T22:47:00Z",
+            "hostname": "worker-2",  # Sensitive info that should not be exposed
+        }
+        self.scheduled_change.failure_reason = json.dumps(failure_context)
+
+        expected = "Database connection lost (failed after 5 attempts)"
```

Comment:
not sure if this is a thing we track; but could be cool to show the `max_retries` as the denominator here; just in case folks are curious (they probably won't be, but it doesn't really cost that many pixels to show something like `"Database connection lost (failed after 5/5 attempts)"`.

That said, we say something about retrying if there are `max_retries` remaining, so maybe I'm overthinking this and it's cleaner this way. Your call.

Reply:
I dig it. I think I'll make it say "(failed after 5 out of 5 attempts)"